File: 2021/golang/day11.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func flashOctopus(G map[point]int, p point, C int, R int) {
	if C != R {
		defer fmt.Println("assert error, C != R")
		os.Exit(1)
	}
	DX := []int{-1, 0, 1}
	DY := []int{-1, 0, 1}
	G[p] = -1
	for _, dx := range DX {
		for _, dy := range DY {
			nextP := point{x: p.x + dx, y: p.y + dy}
			if p.x+dx >= 0 && p.x+dx < C && p.y+dy >= 0 && p.y+dy < R && G[nextP] != -1 {
				G[nextP]++
				if G[nextP] > 9 {
					flashOctopus(G, nextP, C, R)
				}
			}
		}
	}
}

func Day11() {
	file, _ := os.Open("../input/day11.input")
	defer file.Close()
	G := make(map[point]int)
	scanner := bufio.NewScanner(file)
	R, C, steps, totalFlashes := 0, 0, 0, 0

	for scanner.Scan() {
		line := scanner.Text()
		R++
		C = len(line)
		for x, c := range strings.Split(line, "") {
			n, _ := strconv.Atoi(c)
			G[point{x: x, y: R - 1}] = n
		}
	}

	for {
		for k := range G {
			G[k]++
		}

		for k, v := range G {
			if v > 9 {
				flashOctopus(G, k, R, C)
			}
		}

		lights := 0
		for k, v := range G {
			if v == -1 {
				G[k] = 0
				lights++
			}
		}

		steps++
		if len(G) == lights {
			fmt.Println("Answer B:", steps)
			break
		}
		totalFlashes += lights
		if steps == 100 {
			fmt.Println("Answer A:", totalFlashes)
		}
	}
}

File: 2021/python/day11.py
lines = [i.strip() for i in open("../input/day11.input", encoding="utf-8").readlines()]
steps, A, C, R = 0, 0, len(lines[0]), len(lines)
deltas = [(0, 1), (0, -1), (1, 0), (-1, 0), (1, 1), (1, -1), (-1, 1), (-1, -1)]
G = {(col, row): int(lines[row][col]) for col in range(C) for row in range(R)}

def do_increment(pos, hasflashed, S):
    G[pos] += 1
    if G[pos] > 9:
        S.add(pos)
        hasflashed.add(pos)
        G[pos] = 0
        return 1
    return 0

while True:
    steps += 1
    B, hasflashed, S = 0, set(), set()
    for k in G.keys():
        B += do_increment(k, hasflashed, S)

    while S:
        pos = S.pop()
        for d in deltas:
            c, r = pos[0] + d[0], pos[1] + d[1]
            if 0 <= c < C and 0 <= r < R and (c, r) not in hasflashed:
                B += do_increment((c, r), hasflashed, S)
    A += B

    if steps == 100:
        print("Answer A:", A)
    if B == C * R:
        print("Answer B", steps)
        break
